pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c

```diff
@@ -145,6 +145,8 @@ int main(gint argc, gchar **argv) {
   else {
     int status;
 
+    fail_if(signal(SIGINT, SIG_IGN) == SIG_ERR);
```

Comment:
I would rather make the child the forground process like this: tcsetpgrp in the child and block SIGTTOU signal in the parent. Than all signals are only forwarded to the child.

Reply:
I played around with this a bit, but I couldn't find a way to do it that doesn't break `^Z`.

Reply:
Did `^Z` work before?

Reply:
It seems to work for me, with or without this PR, under zsh.

Reply:
The issue so is that you can no longer stop chrootenv with ctrl-c, which also seems a problem.

Reply:
setpgid might be also required: https://www.gnu.org/software/libc/manual/html_node/Launching-Jobs.html

Reply:
Yeah, I needed `setpgid`, but I still couldn't get ^Z to work.

As for ^C, is there any reason chrootenv should deal with it?  It seems like if it's just a wrapper, the child process should be responsible.

Reply:
It's not like it should deal with it. But it should be the child that receive those rather than the parent - Is this still the case
The way I read the code is that the parent is ignoring sigint and the child won't receive it.

Here is an alternative approach that is used in systemd:
https://github.com/systemd/systemd-stable/commit/6f9c8af500fd4362d656917703eab4f941e7e45a#diff-986d1d7b30d37ed34cd2d166891190eb817018b0a2e485e0b455486c56746f31R58

parent does ` ioctl(STDIN_FILENO, TIOCNOTTY)` and child does `setsid`. According to the comments this seems to be cleaner? Regarding ctrl-z: from my understanding it should send SIGSTOP to the process that controls the terminal. Than waitpid would return WIFSIGNALED(status) == 1 and the wrapper would do a SIGSTOP on its own pid.
Here is how I handle that in cntr: https://github.com/Mic92/cntr/blob/b529dc67e27a3e1840afa6e65fd404c97e8b7321/src/attach/parent.rs#L39
This might be the more correct version because it also does SIGCONT after resuming.